source/s2n/tls_channel_handler.c

```diff
@@ -0,0 +1,738 @@
+}
+
+static void s_run_read(struct aws_task *task, void *arg, aws_task_status status) {
+    task->fn = NULL;
+    task->arg = NULL;
+
+    if (status == AWS_TASK_STATUS_RUN_READY) {
+        struct aws_channel_handler *handler = (struct aws_channel_handler *)arg;
+        struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+        s_s2n_handler_process_read_message(handler, s2n_handler->slot, NULL);
+    }
+}
+
+static int s_s2n_handler_increment_read_window(
+    struct aws_channel_handler *handler,
+    struct aws_channel_slot *slot,
+    size_t size) {
+    (void)size;
+    struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+
+    size_t downstream_size = aws_channel_slot_downstream_read_window(slot);
+    size_t current_window_size = slot->window_size;
+
+    if (downstream_size <= current_window_size) {
+        size_t likely_records_count = (downstream_size - current_window_size) % MAX_RECORD_SIZE;
+        size_t offset_size = likely_records_count * (EST_TLS_RECORD_OVERHEAD);
+        size_t window_update_size = (downstream_size - current_window_size) + offset_size;
+        aws_channel_slot_increment_read_window(slot, window_update_size);
+    }
+
+    if (s2n_handler->negotiation_finished) {
+        /* we have messages in a queue and they need to be run after the socket has popped (even if it didn't have data
+         * to read). Alternatively, s2n reads entire records at a time, so we'll need to grab whatever we can and we
+         * have no idea what's going on inside there. So we need to attempt another read.*/
+        s2n_handler->sequential_tasks.fn = s_run_read;
+        s2n_handler->sequential_tasks.arg = handler;
+        aws_channel_schedule_task_now(slot->channel, &s2n_handler->sequential_tasks);
+    }
+
+    return AWS_OP_SUCCESS;
+}
+
+static size_t s_s2n_handler_initial_window_size(struct aws_channel_handler *handler) {
+    (void)handler;
+
+    return EST_HANDSHAKE_SIZE;
+}
+
+struct aws_byte_buf aws_tls_handler_protocol(struct aws_channel_handler *handler) {
+    struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+    return s2n_handler->protocol;
+}
+
+struct aws_byte_buf aws_tls_handler_server_name(struct aws_channel_handler *handler) {
+    struct s2n_handler *s2n_handler = (struct s2n_handler *)handler->impl;
+    return s2n_handler->server_name;
+}
+
+static struct aws_channel_handler_vtable s_handler_vtable = {
+    .destroy = s_s2n_handler_destroy,
+    .process_read_message = s_s2n_handler_process_read_message,
+    .process_write_message = s_s2n_handler_process_write_message,
+    .shutdown = s_s2n_handler_shutdown,
+    .increment_read_window = s_s2n_handler_increment_read_window,
+    .initial_window_size = s_s2n_handler_initial_window_size,
+};
+
+static int s_parse_protocol_preferences(
+    const char *alpn_list_str,
+    const char protocol_output[4][128],
```

Comment:
4 is probably enough, but is there any reason this is hardcoded?

Reply:
yeah, save on heap allocations